### PR TITLE
ARGV cleanup: Deprecate ARGV methods that are never used internally

### PR DESCRIPTION
**Pull request by BenMusch**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Addresses part of https://github.com/Homebrew/brew/issues/5730

This PR deletes `ARGV` methods which are never used internally in `ARGV` (for example, it keeps `build_head?` because it is used in `build_stable?`, and refactoring methods which have internal usage is a bit more complex of a question in terms of how it is refactored. Will definitely look into cleaning that up later, but wanted to open this PR for all the work which could be done without too much thought

### CODE REVIEW

File: Library/Homebrew/test/formula_installer_spec.rb [hunk 1]
```diff
@@ -87,10 +87,8 @@ def temporary_install(formula)
   end
 
   specify "Formula is not poured from bottle when compiler specified" do
-    expect(ARGV.cc).to be nil
-
-    cc_arg = "--cc=clang"
-    ARGV << cc_arg
+    expect(Homebrew.args.cc).to be nil
+    allow(Homebrew.args).to receive(:cc).and_return("clang")
```
Comment (BenMusch) [inline Library/Homebrew/test/formula_installer_spec.rb]: Just wanna make sure: This is a valid replacement in tests for adding values to `ARGV`, right?

File: Library/Homebrew/test/sandbox_spec.rb [hunk 2]
```diff
@@ -47,7 +48,7 @@
     end
 
     it "ignores bogus Python error" do
-      ENV["HOMEBREW_VERBOSE"] = "1"
```
Comment (BenMusch) [inline Library/Homebrew/test/sandbox_spec.rb]: Same here: Wanna make sure it's okay to replace env var values with Homebrew.args stubs

**Comment by BenMusch** [comment]
Not really sure I follow why tests are failing (`brew tests` passes locally), but it definitely seems like it may be related to my changes. Any idea what's going on?

File: Library/Homebrew/test/formula_installer_spec.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/formula_installer_spec.rb]: Yep 👍

File: Library/Homebrew/test/sandbox_spec.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/sandbox_spec.rb]: Yep 👍

**Comment by MikeMcQuaid** [comment]
> Looks like brew test bot might have to be updated? https://github.com/Homebrew/homebrew-test-bot/blob/master/cmd/brew-test-bot.rb#L274 If so, are there any other packages this might affect?

Good catch! The following will likely need updates:
https://github.com/Homebrew/homebrew-test-bot
https://github.com/Homebrew/homebrew-services
https://github.com/Homebrew/homebrew-bundle

**Comment by scpeters** [comment]
Also https://github.com/Homebrew/homebrew-livecheck

Perhaps we could deprecate these old methods to give downstream taps a chance to notice the changes (see https://github.com/Homebrew/brew/pull/5477 for an example). Then we could merge this without being blocked or breaking homebrew commands in external taps.

**Comment by MikeMcQuaid** [comment]
> Perhaps we could deprecate these old methods to give downstream taps a chance to notice the changes (see #5477 for an example).

This seems like a good call 👍 . For those that we notice have been used anywhere outside Homebrew/brew an `odeprecated` and `odisabled` make sense before their removal. For consistency they could also call back to `Homebrew.args.foo` rather than querying `ARGV.foo` directly.

**Comment by BenMusch** [comment]
Looks like the deprecations are still failing the build -- should I be doing this differently? https://dev.azure.com/Homebrew/Homebrew/_build/results?buildId=4180

**Comment by jonchang** [comment]
> Looks like the deprecations are still failing the build -- should I be doing this differently? https://dev.azure.com/Homebrew/Homebrew/_build/results?buildId=4180

The issue is in https://github.com/Homebrew/homebrew-test-bot/, not your pull request.

I believe you'll have to do the pull request in several parts: first to add the new API, then a pull request to test-bot to use the new API, then the deprecations in a final pull request to deprecate the old API. Probably a pull request to homebrew-core fixing all uses of the old API should be in there somewhere as well.

Longer term we should probably look at moving the homebrew/test-bot code into homebrew/brew.

**Comment by jonchang** [comment]
Actually it looks like you're not adding any new APIs here, just deprecating old ones. So a pull request to homebrew-test-bot fixing its calls should be sufficient.

**Comment by BenMusch** [comment]
Ah thanks @jonchang! Didn't see the line about throwing an error when `homebrew_developer?` is true. Opened a PR to change the test-bot: https://github.com/Homebrew/homebrew-test-bot/pull/238

**Comment by MikeMcQuaid** [comment]
@BenMusch Some failing tests but not `brew test-bot` any more 🎉

**Comment by MikeMcQuaid** [comment]
Cross-posting from https://github.com/Homebrew/homebrew-test-bot/pull/240:

> I would suggest making use of `ARGV.freeze` to catch these cases (both here and in Homebrew/brew) and fix/refactor them before any more `ARGV` to `Homebrew.args` conversion is done.

**Comment by BenMusch** [comment]
@MikeMcQuaid should we also find a way to freeze `ENV`, since Homebrew.args values are also controlled by that?

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid should we also find a way to prevent modification of `ENV` (hopefully without interfering with anything non-argv-related values), since Homebrew.args values are also controlled by that?

@BenMusch Good question. I don't believe there are (m)any cases where `ENV` is overridden and the expectation is that `ARGV.whatever?` will change so we're fine from that perspective. Also, we relatively often need to modify `ENV` for subprocesses so I think we're left with what we have for now (without introducing a new API for ENV which also updates CLI::Parser which I think would be overkill for now).

**Comment by BenMusch** [comment]
Sounds good. Just posting this as a heads up: probably won't be able to make progress on the PR for a week or so due to some schoolwork that's come up. But definitely intend on getting back to it once I'm free

**Comment by MikeMcQuaid** [comment]
No worries, thanks @BenMusch!

**Comment by BenMusch** [comment]
Just a heads up: Busier with schoolwork than I thought so it might not be a while until I pick this up again. Will check-in periodically to keep the PR up-to-date with `master`, and anyone who wants to take over the work on this should feel welcome to if my timeliness becomes an issue!

**Comment by MikeMcQuaid** [comment]
@BenMusch Would you be able to extract the minimal version that still passes tests (perhaps into a new PR)? Thanks!

**Comment by BenMusch** [comment]
@MikeMcQuaid Definitely, would a minimal version of this be: not deprecating existing methods, and only changing the `ARGV.*` to `Homebrew.args.*` when it doesn't result in test failures?

**Comment by scpeters** [comment]
> would a minimal version of this be: not deprecating existing methods, and only changing the ARGV.* to Homebrew.args.* when it doesn't result in test failures?

that sounds good to me

**Pull request closed without merging**